### ISSUE TITLE
Support `x-apollo-tracing` HTTP header to passthrough tracing extensions

### ISSUE THREAD

**Issue by schickling** [apollographql/apollo-engine-js#153]
To my understanding, Apollo Engine doesn't passthrough tracing extensions by default (which makes sense). However, there are usecases where you (as a client) would like to receive the "original" extensions. For tracing, this kind of behaviour is implemented by `graphql-yoga` when a client sets the `x-apollo-tracing` (see [code](https://github.com/graphcool/graphql-yoga/blob/master/src/index.ts#L132)).

It would be great if this would also be supported by Apollo Engine by default.

**Comment by martijnwalraven** [comment]
This isn't currently documented, but Apollo Engine already supports a set of extensions to pass-through under an `includeExtensions` key in a request's extensions. Including `tracing` is blacklisted by default however, because it's considered a security risk. I don't think the Engine config supports changing the set of blacklisted extensions, but that seems like something we could expose. You probably don't want to do this in production however.

The reason for relying on request extensions instead of an HTTP header is that we want the mechanism to be transport agnostic. This way, it should also work with a WebSocket transport for example (once we add support for that to Engine).

**Comment by schickling** [comment]
Interesting! Can you send over a `curl` example how this would be enabled on a per request basis?

**Comment by glasser** [comment]
Right, we currently document how to adjust the "default strip" and the "blacklist" list but not how to actually change the request from the client.

To use it, include an includeInResponse extension, after setting `frontends.extensions.blacklist` to the empty list (or some list containing something other than :

```
curl http://localhost:4000/graphql -H 'Content-type: application/json' -d '{"query": "query x { posts { id } }", "extensions": {"includeInResponse":["tracing"]}}'
```

Note that apollo-client currently requires you to set the (undocumented?) includeExtensions option to send client->server extensions, and there's some weirdness about combining this with its main user apollo-link-persisted-queries (which can override includeExtensions to 'false' if used with a server that doesn't support PQs). We'll clean this up when we fully support/document client->server extensions.

**Comment by schickling** [comment]
Thanks! I wasn't aware the `extensions` is also supported for GraphQL requests. Can you point me to the according part in the spec?

**Comment by glasser** [comment]
client->server extensions is an Apollo extension that we've been trialling with apollo-link-persisted-queries to learn more about how it works best before proposing it more formally for GraphQL in general. (For example, we'll definitely want to clean up the includeExtensions API mentioned above before we do so.)

It's basically the exact analogue of server->client extensions: an extra map in the request instead of in the response. There's definitely some interesting questions to figure out around "what if you don't want the server to process your message unless it understands your extensions", etc.
